File: hugegraph-client/src/test/java/org/apache/hugegraph/unit/PropertyKeyTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with this
 * work for additional information regarding copyright ownership. The ASF
 * licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package org.apache.hugegraph.unit;

import org.apache.hugegraph.structure.constant.Cardinality;
import org.apache.hugegraph.structure.constant.DataType;
import org.apache.hugegraph.structure.constant.HugeType;
import org.apache.hugegraph.structure.schema.PropertyKey;
import org.apache.hugegraph.testutil.Assert;
import org.junit.Test;

public class PropertyKeyTest {

    @Test
    public void testPropertyKey() {
        PropertyKey.Builder builder = new PropertyKey.BuilderImpl("name",
                                                                  null);
        PropertyKey propertyKey = builder.dataType(DataType.INT)
                                         .cardinality(Cardinality.SINGLE)
                                         .userdata("min", 1)
                                         .userdata("max", 100)
                                         .build();

        String pkString = "{name=name, cardinality=SINGLE, dataType=INT, " +
                          "aggregateType=NONE, properties=[], " +
                          "writeType=OLTP}";
        Assert.assertEquals(pkString, propertyKey.toString());
        Assert.assertEquals(HugeType.PROPERTY_KEY.string(), propertyKey.type());
        Assert.assertEquals(0, propertyKey.aggregateType().code());
        Assert.assertEquals("none", propertyKey.aggregateType().string());
    }

    @Test
    public void testPropertyKeyV46() {
        PropertyKey.Builder builder = new PropertyKey.BuilderImpl("name",
                                                                  null);
        PropertyKey propertyKey = builder.dataType(DataType.INT)
                                         .cardinality(Cardinality.SINGLE)
                                         .userdata("min", 1)
                                         .userdata("max", 100)
                                         .build();

        PropertyKey.PropertyKeyV46 propertyKeyV46 = propertyKey.switchV46();
        String pkV46String = "{name=name, cardinality=SINGLE, " +
                             "dataType=INT, properties=[]}";
        Assert.assertEquals(pkV46String, propertyKeyV46.toString());
        Assert.assertEquals(HugeType.PROPERTY_KEY.string(),
                            propertyKeyV46.type());
    }

    @Test
    public void testPropertyKeyV58() {
        PropertyKey.Builder builder = new PropertyKey.BuilderImpl("name",
                                                                  null);
        PropertyKey propertyKey = builder.dataType(DataType.INT)
                                         .cardinality(Cardinality.SINGLE)
                                         .userdata("min", 1)
                                         .userdata("max", 100)
                                         .build();

        PropertyKey.PropertyKeyV58 propertyKeyV58 = propertyKey.switchV58();
        String pkV58String = "{name=name, cardinality=SINGLE, " +
                             "dataType=INT, aggregateType=NONE, properties=[]}";
        Assert.assertEquals(pkV58String, propertyKeyV58.toString());
        Assert.assertEquals(HugeType.PROPERTY_KEY.string(),
                            propertyKeyV58.type());
    }
}

File: hugegraph-client-go/api/v1/propertykey/propertykey_test.go
/*
* Licensed to the Apache Software Foundation (ASF) under one or more
* contributor license agreements. See the NOTICE file distributed with this
* work for additional information regarding copyright ownership. The ASF
* licenses this file to You under the Apache License, Version 2.0 (the
* "License"); you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
* License for the specific language governing permissions and limitations
* under the License.
 */

package propertykey_test

import (
    "fmt"
    "github.com/apache/incubator-hugegraph-toolchain/hugegraph-client-go"
    "github.com/apache/incubator-hugegraph-toolchain/hugegraph-client-go/api/v1/propertykey"
    "github.com/apache/incubator-hugegraph-toolchain/hugegraph-client-go/internal/model"
    "log"
    "math/rand"
    "testing"
    "time"
)

func TestPropertyKey(t *testing.T) {

    client, err := hugegraph.NewDefaultCommonClient()
    if err != nil {
        log.Println(err)
    }
    rand.Seed(time.Now().UnixNano())
    name := fmt.Sprintf("testProperty%d", rand.Intn(99999))

    // create propertyKey
    respCreate, err := client.Propertykey.Create(
        client.Propertykey.Create.WithReqData(
            propertykey.CreateRequestData{
                Name:        name,
                DataType:    model.PropertyDataTypeInt,
                Cardinality: model.PropertyCardinalitySingle,
            },
        ),
    )
    if err != nil {
        t.Errorf(err.Error())
    }
    if respCreate.Data.PropertyKey.Name != name || respCreate.Data.PropertyKey.ID <= 0 {
        t.Errorf("create propertyKey failed")
    }

    // propertyKey get all
    respGetAll, err := client.Propertykey.GetAll()
    if err != nil {
        t.Errorf(err.Error())
    }

    hasCreated := false
    for _, pk := range respGetAll.Data.Propertykeys {
        if pk.Name == name {
            hasCreated = true
        }
    }
    if !hasCreated {
        t.Errorf("get all propertyKey failed")
    }

    // propertyKey update user_data
    respUpdateUserdata, err := client.Propertykey.UpdateUserdata(
        client.Propertykey.UpdateUserdata.WithReqData(
            propertykey.UpdateUserdataRequestData{
                Action: model.ActionAppend,
                Name:   name,
                UserData: struct {
                    Min int `json:"min"`
                    Max int `json:"max"`
                }(struct {
                    Min int
                    Max int
                }{
                    Min: 1,
                    Max: 10,
                }),
            },
        ),
    )
    if err != nil {
        t.Errorf(err.Error())
    }
    if respUpdateUserdata.Data.PropertyKey.UserData.Max != 10 {
        t.Errorf("update userdata propertyKey failed")
    }

    // propertyKey get by name
    respGetByName, err := client.Propertykey.GetByName(
        client.Propertykey.GetByName.WithName(name),
    )
    if err != nil {
        t.Errorf(err.Error())
    }
    if respGetByName.Data.Name != name || respGetByName.Data.ID <= 0 || respGetByName.Data.UserData.Max != 10 {
        t.Errorf("getByName propertyKey failed")
    }

    // propertyKey delete
    respDelete, err := client.Propertykey.DeleteByName(
        client.Propertykey.DeleteByName.WithName(name),
    )
    if err != nil {
        t.Errorf(err.Error())
    }
    if respDelete.StatusCode > 299 {
        t.Errorf("delete propertyKey failed")
    }
}
